File: Gaussian_HMC.cpp
#include <iostream>
#include <cmath>
#include<fstream>
const int niter=1000;
const int ntau=10;
const double dtau=0.1e0;
/******************************************************************/
/*** Gaussian Random Number Generator with Box Muller Algorithm ***/
/******************************************************************/
int BoxMuller(double& p, double& q){
  
  double pi=2e0*asin(1e0);
  //uniform random numbers between 0 and 1
  double r = (double)rand()/RAND_MAX;
  double s = (double)rand()/RAND_MAX;
  //Gaussian random numbers, 
  //with weights proportional to e^{-p^2/2} and e^{-q^2/2}
  p=sqrt(-2e0*log(r))*sin(2e0*pi*s);
  q=sqrt(-2e0*log(r))*cos(2e0*pi*s);
  
  return 0;
}
/*********************************/
/*** Calculation of the action ***/
/*********************************/
// When you change the action, you should also change dH/dx,
// specified in "calc_delh". 
double calc_action(const double x){
  
  double action=0.5e0*x*x;
  
  return action;
}
/**************************************/
/*** Calculation of the Hamiltonian ***/
/**************************************/
double calc_hamiltonian(const double x,const double p){
  
  double ham=calc_action(x);
  
  ham=ham+0.5e0*p*p;
  
  return ham;
}
/****************************/
/*** Calculation of dH/dx ***/
/****************************/
// Derivative of the Hamiltonian with respect to x, 
// which is equivalent to the derivative of the action.
// When you change "calc_action", you have to change this part as well. 
double calc_delh(const double x){
  
  double delh=x;
  
  return delh;
}
/***************************/
/*** Molecular evolution ***/
/***************************/
int Molecular_Dynamics(double& x,double& ham_init,double& ham_fin){

  double r1,r2;
  BoxMuller(r1,r2);
  double p=r1;
  
  //*** calculate Hamiltonian ***
  ham_init=calc_hamiltonian(x,p);
  //*** first step of leap frog ***
  x=x+p*0.5e0*dtau;
  //*** 2nd, ..., Ntau-th steps ***
  for(int step=1; step!=ntau; step++){    
    double delh=calc_delh(x);
    p=p-delh*dtau;
    x=x+p*dtau;
  }
  //*** last step of leap frog ***
  double delh=calc_delh(x);
  p=p-delh*dtau;
  x=x+p*0.5e0*dtau;
  //*** calculate Hamiltonian again ***
  ham_fin=calc_hamiltonian(x,p);
  
  return 0;
}

int main()
{ 
  srand((unsigned)time(NULL));
  /*********************************/
  /* Set the initial configuration */
  /*********************************/
  double x=0e0;
  /*****************/
  /*** Main part ***/
  /*****************/
  std::ofstream outputfile("output.txt");
  int naccept=0;//counter for the number of acceptance
  double sum_xx=0e0;//sum of x^2, useed for <x^2>
  
  for(int iter=0; iter!=niter; iter++){
    
    double backup_x=x;
    double ham_init,ham_fin;
    Molecular_Dynamics(x,ham_init,ham_fin);
    double metropolis = (double)rand()/RAND_MAX;
    if(exp(ham_init-ham_fin) > metropolis){
      //accept
      naccept=naccept+1;
    }else{
      //reject
      x=backup_x;
    }
    
    /*******************/
    /*** data output ***/
    /*******************/
    sum_xx=sum_xx+x*x;

    // output x, <x^2>, acceptance
    
    std::cout << std::fixed << std::setprecision(6) 
      << x << "   " 
      << sum_xx/((double)(iter+1)) <<  "   " 
      << ((double)naccept)/((double)iter+1) 
      << std::endl;
     
    outputfile << std::fixed << std::setprecision(6) 
      << x << "   " 
      << sum_xx/((double)(iter+1)) <<  "   " 
      << ((double)naccept)/((double)iter+1) 
      << std::endl;
  }
    
  outputfile.close();
  
  return 0;
}
  

File: Gaussian_HMC.py
import numpy as np
from matplotlib import pyplot as plt
# seed of random numbers is the system time by default
# Numpy has Gaussian random number generator, so we don't use Box-Muller

NITER = 500000
NTAU = 10
DTAU = 0.1
##################################
### Calculation of the action  ###
##################################
#When you change the action, you should also change dH/dx, specified in "calc_delh".
def calc_action(x):
    return 0.5*x*x
######################################
### Calculation of the Hamiltonian ###
######################################
def calc_hamiltonian(x,p):
    ham = calc_action(x)
    ham += 0.5*p*p
    return ham
#############################
### Calculation of dH/dx  ###
#############################
def calc_delh(x):
    delh = x
    return delh
#############################
### Molecular evolution  ###
#############################
def Molecular_Dynamics(x):
    # Momentum p is chosen to be Gaussian ##
    p = np.random.randn()
    # calculate Hamiltonian ##
    ham_init = calc_hamiltonian(x,p)
    # first step of leap frog ##
    x += p*0.5*DTAU
    # 2nd, ..., Ntau-th steps ##
    for step in range(1,NTAU):
        delh = calc_delh(x)
        p += -delh*DTAU
        x += p*DTAU
        
    # last step of leap frog ##
    delh = calc_delh(x)
    p += -delh*DTAU
    x += p*0.5e0*DTAU

    # calculate Hamiltonian again ##
    ham_fin=calc_hamiltonian(x,p)
    return x,ham_init,ham_fin
#####################################
### Set the initial configuration ###
#####################################
x = 0
naccept = 0
sum_xx = 0.0 #sum of x^2, useed for <x^2>

# data for plot
data_for_plot = []

#################
### Main Loop ###
#################
for iter in range(NITER):
    backup_x=x
    x,ham_init,ham_fin = Molecular_Dynamics(x)
    metropolis = np.random.uniform(0,1)
    if(np.exp(ham_init-ham_fin) > metropolis):#accept
        naccept += 1
    else:#reject
        x = backup_x
    ###################
    ### data output ###
    ###################
    sum_xx=sum_xx+x*x
    # output x, <x^2>, acceptance
    print(x,sum_xx/(iter+1),naccept/(iter+1))

    # save configuration for plot
    data_for_plot.append(x)

###########################################################
### plot the histogram and compare with analytic result ###
###########################################################
def gauss(x):
    return 1.0/np.sqrt(2.0*np.pi)*np.exp( -x*x/2.0 )

y=np.arange(-4.0,4.0,0.1)
plt.xlabel("$x$")
plt.ylabel("$P(x)$")
plt.plot(y,gauss(y), label="exact")
plt.hist(data_for_plot,bins=80,density=True, label="MCMC")
plt.legend()
plt.show() 
